lmdeploy/pytorch_poc/models/falcon.py

```diff
@@ -0,0 +1,361 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Adapter from:
+# https://huggingface.co/tiiuae/falcon-7b-instruct
+# https://github.com/huggingface/transformers/blob/v4.33-release/src/transformers/models/falcon/modeling_falcon.py  # noqa
+
+from typing import Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+import torch.utils.checkpoint
+from transformers.modeling_outputs import \
+    BaseModelOutputWithPastAndCrossAttentions
+from transformers.models.falcon.modeling_falcon import build_alibi_tensor
+
+from lmdeploy.pytorch_poc.kernels import (alibi_paged_attention_fwd,
+                                          paged_attention_fwd)
+
+
+# rotary pos emb helpers
+# (torch.jit.script does not seem to support staticmethod...)
+def rotate_half(x):
+    x1, x2 = x[..., :x.shape[-1] // 2], x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+class PatchedFalconRotaryEmbedding(nn.Module):
+    """Implementation adapted from Huggingface transformers."""
+
+    def _patched_set_cos_sin_cache(self, seq_len, device, dtype):
+        self.seq_len_cached = seq_len
+        t = torch.arange(seq_len, device=device, dtype=self.inv_freq.dtype)
+        freqs = torch.einsum('i,j->ij', t, self.inv_freq)
+        emb = torch.cat((freqs, freqs), dim=-1).to(device)
+
+        if dtype in [torch.float16, torch.bfloat16]:
+            emb = emb.float()
+
+        self.cos_cached = emb.cos()[None, :, :]
+        self.sin_cached = emb.sin()[None, :, :]
+
+        self.cos_cached = self.cos_cached.type(dtype)
+        self.sin_cached = self.sin_cached.type(dtype)
+
+    def patched_cos_sin(self,
+                        position_ids: torch.Tensor,
+                        device='cpu',
+                        dtype=torch.bfloat16) -> torch.Tensor:
+        total_length = int(position_ids.max().item()) + 1
+
+        if (self.seq_len_cached is None) or (total_length >
+                                             self.seq_len_cached):
+            self._patched_set_cos_sin_cache(total_length, device, dtype)
+        # position_ids.shape == [1, packed_seq_len]
+        return (
+            self.cos_cached[:, position_ids[0], None, :],
```

Comment:
position_ids has not been continuous yet. This embedding might not be able to do the batched inference.
I have added a position_ids_1d in the context.

Reply:
Should I just continuous position_ids?

Reply:
Updated to use position1d